sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/SingletonPatternCheck.kt

```diff
@@ -0,0 +1,131 @@
+        }
+    }
+}
+
+private class SingletonClassCandidateExtractor : KtTreeVisitor() {
+
+    val singletonClassCandidates = mutableSetOf<String>()
+
+    override fun visitClass(klass: KtClass) {
+        if (isSingletonClassCandidate(klass)) {
+            klass.fqName?.let {
+                singletonClassCandidates.add(it.toString())
+            }
+        }
+    }
+}
+
+private fun isSingletonClassCandidate(klass: KtClass): Boolean {
+    val constructors = klass.allConstructors
+    return klass.isPrivate() || (constructors.isNotEmpty() && constructors.all(KtModifierListOwner::isPrivate))
+}
+
+private fun KtModifierListOwner.isPrivate(): Boolean = visibilityModifier()?.text == "private"
+
+private class SingleConstructorCallExtractor(
+    private val singletonClassCandidates: MutableSet<String>,
+    private val bindingContext: BindingContext,
+) : KtTreeVisitor() {
+
+    val singleConstructorCallByClass: MutableMap<String, KtCallExpression> = mutableMapOf()
+
+    override fun visitCallExpression(expression: KtCallExpression) {
+        val fqName = getCalledConstructorClassFqNameOrNull(expression) ?: return
+        if (!singletonClassCandidates.contains(fqName)) return
+        if (singleConstructorCallByClass.put(fqName, expression) != null) {
+            singleConstructorCallByClass.remove(fqName)
+            singletonClassCandidates.remove(fqName)
+        }
+    }
+
+    private fun getCalledConstructorClassFqNameOrNull(expression: KtCallExpression): String? {
+        val calleeReference = expression.calleeExpression as? KtReferenceExpression ?: return null
+        val calledConstructor = bindingContext[BindingContext.REFERENCE_TARGET, calleeReference] as? ConstructorDescriptor
+        return calledConstructor?.constructedClass?.fqNameSafe?.toString()
+    }
+}
+
+private fun isInitializingCall(callExpression: KtCallExpression): Boolean {
+    val property = callExpression.parent as? KtProperty ?: return false
+    return isDeclaredInCompanionObject(property)
+}
+
+private fun isLazyInitializingCall(callExpression: KtCallExpression, bindingContext: BindingContext): Boolean {
+    val bodyExpression = callExpression.parent as? KtBlockExpression ?: return false
+    if (bodyExpression.statements.last() != callExpression) return false
+    val call = (((bodyExpression.parent as? KtFunctionLiteral)
+        ?.parent as? KtLambdaExpression)
+        ?.parent as? KtLambdaArgument)
+        ?.parent as? KtCallExpression ?: return false
+
+    val property = (call.parent as? KtPropertyDelegate)?.parent as? KtProperty ?: return false
+    if (call.valueArguments.size != 1 || !isDeclaredInCompanionObject(property)) return false
+    val calleeReference = call.calleeExpression as? KtReferenceExpression ?: return false
+    return bindingContext[BindingContext.REFERENCE_TARGET, calleeReference]?.name?.toString() == "lazy"
```

Comment:
This logic should be replaceable by some generic logic we already have. For one, there is `getResolvedCall(BindingContext)` defined on e.g. the call expression (amongst others). Even better, you can use a FunMatcher here as well.

Reply:
Fixed. Didn't use a FunMatcher though, because here I do not want to match the function, but get the constructor type name, so I can differentiate between three states of the class:

- instantiated never
- instantiated exactly once
- instantiated more than once

I do realize though that here is a bit code duplication with the function matcher, since it determines the constructed class name in exactly the same way. Should we leverage this and create some helper functions here that could be reused, wdyt?